sklearn/base.py

```diff
@@ -410,7 +418,7 @@ def _validate_data(self, X, y=None, reset=True,
         """
 
         if y is None:
-            if self._get_tags()['requires_y']:
+            if reset and self._get_tags()['requires_y']:
```

Comment:
this means that if `y` isn't passed the second time `partial_fit` is called, then the error won't be properly raised. Do we want that?

Reply:
This kind of makes this inconvenient for calling in non-fit methods. For predict, we would set `reset=False` and not provide a `y`. This if statement almost assumes that `_validate_data` is only called in `fit` or `partial_fit`.

Reply:
We spoke about it here: https://github.com/scikit-learn/scikit-learn/pull/18001

Reply:
A solution would be to add a `requires_y` kwargs to `_validate_data`:

```py
    def _validate_data(self, X, y=None, reset=True, requires_y='auto',
                       validate_separately=False, **check_params):
```

Where `requires_y='auto'` means using the tag and it can also be a bool.

Reply:
IIRC our chat with @amueller a few months ago, the cleanest solution we had agreed on was to call `_validate_data` only in `fit` and in the first `partial_fit`, and to define another method for the rest (`predict`, `transform`, subsequent `partial_fit`, etc.). This might require changing `_validate_data` a little, which is fine since it's private

Reply:
I'll explore the second method approach in another PR.

Currently, I kind of like this `requires_y='use_tag'` approach, which allows the caller to specify `requires_y` or default to the estimator tag.

Reply:
@NicolasHug What do you think of using `requires_y='use_tag'`, which results in this:

```py
if ((requires_y == 'use_tag' and self._get_tags()['requires_y']) or 
      (isinstance(requires_y, bool) and requires_y)):
```

I am trying to avoid the discussion of "how to name the new method" (like in https://github.com/scikit-learn/enhancement_proposals/pull/22), if there is a simple way to add a kwarg to `_validate_data`.

Reply:
If we add a new parameter to `_validate_data`, isn't the `requires_y` tag redundant now?

Reply:
It also means that we need to pass `requires_y=False` in predict, transform, etc.

So in those methods the call becomes e.g.:

```py
X = self._validate_data(X, dtype=np.float64, requires_y=False)
```

In terms of API, it feels strange to only pass X and still have to say "I'm not passing y"


Could we instead change the default of `y` in `validate_data`? Since `y` defaults to `None` in `fit`, if we made it default to e.g. a custom constant `_NO_Y = '__no_y'` in `validate_data`, we might be able to avoid the extra parameter, or the new method.

Reply:
We could introduce an alias for `self._validate_input_data()` for methods that do not have `y` argument.

Reply:
> We could introduce an alias for self._validate_input_data() for methods that do not have y argument.

If this is suggesting using another method, I am trying to avoid the confusion of deciding which `_validate_*` method to use.

>  Since y defaults to None in fit, if we made it default to e.g. a custom constant _NO_Y = '__no_y' in validate_data, we might be able to avoid the extra parameter, or the new method.

This seems like it would work. This means all `fit` calls must always explicitly pass `y`.

Reply:
For times like these, I wish we had a `_validate_X` and a `_validate_X_y`, then it would be more obvious which to use.

"_validate_data" is fairly ambiguous.

Reply:
> This means all fit calls must always explicitly pass y.

You mean that `y` must be explicitly passed to `fit`, or that `y` must be explicitly passed to `_validate_data` in `fit`?

Either way, I'm not sure I understand the reason.

Reply:
I updated the PR to use the `__NO_OP` placeholder. This means that it will be the callers responsibly to call this with `y` to get the `requires_y` check:

```py
self._validate_data(X, None)  # will check that `y` is consistent with `require_y` tag

self._validate_data(X)  # will ignore `requires_y` tag
```

The alternative is to recommend setting `y` (in `_validate_data`) everywhere in `fit`. In other words:

```py
def fit(self, X, y=None):
    # note that `y` is set explicitly to `None`, because the docstring says `y` is ignored
    X = self._validate_data(X, None)
```

This way the `requires_y` can always be checked.

Reply:
> This means that it will be the callers responsibly to call this with y to get the requires_y check

I think that's fine because the reason we introduced the tag was to give a decent error message when `y` should be passed (e.g. for supervized estimators), but the user left it to None. Without the tag, they would get a weird tuple unpacking error with `X, y = _validate_data(X, y)`. In other words, the tag is only useful when we pass both X and y, so that's fine to ignore it when we just pass X.

Could you remove the use of the `requires_y` parameter? I still see it in a few places ;)

Reply:
I think I am okay with delegating the responsibility of passing `y` to the `_validate_data`'s caller:

- If the caller passes `y`, then `_validate_data` will validate `y`.
- If the caller does not pass `y`, then do nothing (in terms of `y`).

Reply:
> Could you remove the use of the requires_y parameter? I still see it in a few places ;)

All gone now!